rook/main.py

```diff
@@ -0,0 +1,278 @@
+                             cwd=directory,
+                             universal_newlines=True)
+  output = process.communicate()[0]
+  retcode = process.returncode
+  passed = (retcode == 0)
+  if passed:
+    short = "Success"
+  else:
+    short = "Failed"
+  return (passed, short, output)
+
+def get_testers_and_differs(directory):
+  """
+  imports all the testers and differs in a directory
+  returns a dictionary with all the subclasses of Tester.
+  """
+  tester_dict = {}
+  differ_dict = {}
+  os.sys.path.append(directory)
+  for filename in os.listdir(directory):
+    if filename.endswith(".py") and not filename.startswith("__"):
+      module = __import__(filename[:-3]) #[:-3] to remove .py
+      for name, value in module.__dict__.items():
+        #print("Unknown", name, value)
+        if inspect.isclass(value) and value is not Tester\
+           and issubclass(value, Tester):
+          tester_dict[name] = value
+        if inspect.isclass(value) and value is not Differ\
+           and issubclass(value, Differ):
+          differ_dict[name] = value
+
+  return tester_dict, differ_dict
+
+def sec_format(runtime):
+  """
+  Formats the runtime into a string of the number seconds.
+  If runtime is none, format as None!
+  runtime: float or None, runtime to be formated.
+  return str of runtime.
+  """
+  if isinstance(runtime, float):
+    return "{:6.2f}sec".format(runtime)
+  return "  None!  "
+
+def process_result(index, _input_data, output_data):
+  """
+  This is a callback function that Processes the result of a test.
+  index: int, Index into functions list.
+  _input_data: the input data passed to the function
+  output_data: the output data passed to the function
+  """
+  bucket = output_data.bucket
+  process_test_name = test_name_list[index]
+  if bucket == Tester.bucket_success:
+    results["pass"] += 1
+    for postreq in function_postreq.get(process_test_name, []):
+      if postreq in name_to_id:
+        job_id = name_to_id[postreq]
+        print("Enabling", postreq, job_id)
+        run_pool.enable_job(job_id)
+  elif bucket == Tester.bucket_skip:
+    results["skipped"] += 1
+    print(output_data.message)
+  else:
+    results["fail"] += 1
+    failed_list.append(process_test_name)
+    print(output_data.output)
+    print(output_data.message)
+  number_done = sum(results.values())
+  print("({}/{}) {:7s} ({}) {}".format(number_done, len(function_list),
+                                       Tester.get_bucket_name(bucket),
+                                       sec_format(output_data.runtime),
+                                       process_test_name))
+if __name__ == "__main__":
+
+  test_re = re.compile(args.test_re_raw)
+
+  #XXX fixme to find a better way to the tests directory
+
+  this_dir = os.path.abspath(os.path.dirname(__file__))
+  up_one_dir = os.path.dirname(this_dir)
+  base_test_dir = os.path.join(up_one_dir, "tests")
+
+  print(this_dir, base_test_dir)
```

Comment:
Is this debug printing or meant to be user-informing? Either removal or some additional text to explain what's printed might help.

Reply:
Print removed.